encoder_h264:add parameter Cabac Dct8x8 and DeblockFilter

wudping opened the pull request.

lizhong1008 commented:
+1

xuguangxin commented on encoder/vaapiencoder_h264.cpp:
```diff
@@ -1319,9 +1321,10 @@ bool VaapiEncoderH264::addSliceHeaders (const PicturePtr& picture) const
-        sliceParam->slice_beta_offset_div2 = 2;
 
+        sliceParam->disable_deblocking_filter_idc = !m_videoParamAVC.enableDeblockFilter;
+        sliceParam->slice_alpha_c0_offset_div2 = m_videoParamAVC.deblockAlphaOffset;
```
Hi @lizhong1008 & @wudping 
two question:
1. original sliceParam->slice_alpha_c0_offset_div2 default is 2, this code will change it to 0. Do you think this is ok? 
2. slice_alpha_c0_offset_div2 means offset div 2, but we can't get this hint from deblockAlphaOffset' name, Is it OK?

lizhong1008 replied:
Hi @xuguangxin : 
1. Yeah, I also noted it. I didn't raise it because I thought it's ok to set it to be zero. It has a little effect to deblocking stress,  but no necessary relation on encoding quality. My concern is it may effect QA test result. So would better change it to 2.
2. Good catch. Here deblockAlphaOffset should same as FilterOffsetA defined in h264 spec. Should change to sliceParam->slice_alpha_c0_offset_div2 = m_videoParamAVC.deblockAlphaOffset / 2. 
Same to  slice_beta_offset_div2 and deblockBetaOffset.

lizhong1008 replied:
Hi @xuguangxin : 
1. Yeah, I also noted it. I didn't raise it because I think it's ok to set it to be zero. It has a little effect to deblocking stress,   but has no necessary relation on encoding quality. My concern is it may effect QA test result. So would better change it to 2.
2. Good catch. Here deblockAlphaOffset should same as FilterOffsetA defined in h264 spec. Should change to sliceParam->slice_alpha_c0_offset_div2 = m_videoParamAVC.deblockAlphaOffset / 2. 
Same to  slice_beta_offset_div2 and deblockBetaOffset.

lizhong1008 commented on interface/VideoEncoderDefs.h:
```diff
@@ -263,6 +263,11 @@ typedef struct VideoParamsAVC {
+    bool  enableCabac;
+    bool  enableDct8x8;
+    bool  enableDeblockFilter;
+    int32_t deblockAlphaOffset;
```
Would better add comments: // same as FilterOffsetA  defined in h264 spec.

lizhong1008 commented on interface/VideoEncoderDefs.h:
```diff
@@ -263,6 +263,11 @@ typedef struct VideoParamsAVC {
+    bool  enableDct8x8;
+    bool  enableDeblockFilter;
+    int32_t deblockAlphaOffset;
+    int32_t deblockBetaOffset;
```
// same as FilterOffsetB defined in h264 spec.

wudping commented:
thanks Guangxin and LiZhong for your constructed suggestions.

xuguangxin commented on 0001-encoder_h264-add-parameter-Cabac-Dct8x8-and-DeblockF.patch:
```diff
@@ -0,0 +1,97 @@
+From 9c2943b0680b0c2194ba0a27fadf7a34dcafa5d2 Mon Sep 17 00:00:00 2001
```
Hi @wudping 
You falsely add .patch to your commit, please remove it.

xuguangxin commented on encoder/vaapiencoder_h264.cpp:
```diff
@@ -1319,9 +1321,10 @@ bool VaapiEncoderH264::addSliceHeaders (const PicturePtr& picture) const
-        sliceParam->slice_beta_offset_div2 = 2;
 
+        sliceParam->disable_deblocking_filter_idc = !m_videoParamAVC.enableDeblockFilter;
+        sliceParam->slice_alpha_c0_offset_div2 = m_videoParamAVC.deblockAlphaOffset / 2;
```
@wudping and @lizhong1008 
Just checked spec, slice_alpha_c0_offset_div2 is from spec, do we need change m_videoParamAVC.deblockAlphaOffset to sliceAlphaC0OffsetDiv2? Because deblockAlphaOffset is not so meaningful,  2n and 2n+1 will get same result for deblockAlphaOffset

lizhong1008 replied:
I've no comment to use deblockAlphaOffset or sliceAlphaC0OffsetDiv2. But we would better comment how this name maps h264 spec. For example, if sliceAlphaC0OffsetDiv2 used, comments "same as slice_alpha_c0_offset_div2 defined in h264 spec 7.4.3"

wudping replied:
okay, Thank you two. I will add the comments.

xuguangxin commented on interface/VideoEncoderDefs.h:
```diff
@@ -263,6 +263,11 @@ typedef struct VideoParamsAVC {
+    bool  enableCabac;
+    bool  enableDct8x8;
+    bool  enableDeblockFilter;
+    int32_t deblockAlphaOffsetDiv2; //same as slice_alpha_c0_offset_div2 defined in h264 spec 7.4.3
```
it should be int8_t

wudping replied:
ok, Thanks.

xuguangxin commented:
+1

lizhong1008 commented:
-1. 
0001-encoder_h264-add-parameter-Cabac-Dct8x8-and-DeblockF.patch is added in the first commit, then removed. It's not necessary.

lizhong1008 commented:
@wudping  Maybe you can squash these two commits to be one.

wudping commented:
ok, LiZhong.

dongpingx-wu committed:
encoder_h264:add parameter Cabac Dct8x8 and DeblockFilter

add parameters Cabac Dct8x8 and DeblockFilter into 264 encoder;
add deblock_alph_offset_div2, deblock_beta_offset_div2 into slice header,
for DeblockFilter.

Signed-off-by: wdp <dongpingx.wu@intel.com>

xuguangxin commented on 0001-encoder_h264-add-parameter-Cabac-Dct8x8-and-DeblockF.patch:
```diff
@@ -0,0 +1,254 @@
+From 905a0d05a83dfd2f4f0e994ff4cc628dd1d708e1 Mon Sep 17 00:00:00 2001
```
;), maybe you need 
`git commit -a -m "comments"`
to submit your changes.
Hope you can use "git show" before you do any git push.

lizhong1008 commented:
+1

lizhong1008 commented:
merged as 5ab12246b87ca12a0c1b06b7b697f6c311a07c74
